File: glidercalc/stability.py
"""Static stability and pitch/roll trim for a fully submerged glider.

A submerged vehicle is kept upright not by a waterplane (as a surface ship is)
but by the vertical separation between its centre of buoyancy (CB) and centre of
gravity (CG). For stability the **CB must sit above the CG**; the vertical gap
``BG`` is the stability lever.

Coordinate convention
---------------------
* ``x`` : longitudinal, positive forward (towards the nose)
* ``z`` : vertical, positive up

When the vehicle tilts by an angle ``theta`` the weight (down, through CG) and
buoyancy (up, through CB) form a righting couple::

    M_restore = B * BG * sin(theta)      (B ~= W near neutral buoyancy)

Pitch trim by a moving mass
---------------------------
Gliders pitch by sliding an internal mass (often the battery) fore/aft. Moving a
mass ``m_p`` by a distance ``d`` shifts the CG longitudinally by::

    dx_cg = m_p * d / m_total

The vehicle then pitches until the CG is again vertically below the CB::

    tan(theta) = dx_cg / BG

so a larger BG gives a "stiffer", harder-to-trim vehicle, and a smaller BG
trims more easily but is less stable.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Iterable

from . import constants


@dataclass
class MassComponent:
    """A point mass at a longitudinal/vertical location (SI units)."""

    name: str
    mass_kg: float
    x_m: float = 0.0
    z_m: float = 0.0


def center_of_mass(components: Iterable[MassComponent]) -> tuple[float, float, float]:
    """Return ``(total_mass, x_cg, z_cg)`` for a set of point masses."""
    comps = list(components)
    total = sum(c.mass_kg for c in comps)
    if total <= 0:
        raise ValueError("total mass must be positive")
    x = sum(c.mass_kg * c.x_m for c in comps) / total
    z = sum(c.mass_kg * c.z_m for c in comps) / total
    return total, x, z


@dataclass
class StabilityResult:
    """Static-stability and trim summary (SI units; angles in degrees)."""

    total_mass_kg: float
    cg_x_m: float
    cg_z_m: float
    cb_x_m: float
    cb_z_m: float
    bg_m: float
    is_stable: bool
    righting_stiffness_nm_per_rad: float
    longitudinal_offset_m: float
    natural_pitch_deg: float

    def summary(self) -> str:
        verdict = ("STABLE (CB above CG)" if self.is_stable
                   else "UNSTABLE (CG at or above CB - will roll/pitch over)")
        lines = [
            "Stability & trim",
            "----------------",
            f"  Total mass          : {self.total_mass_kg:.3f} kg",
            f"  CG (x, z)           : ({self.cg_x_m:+.4f}, {self.cg_z_m:+.4f}) m",
            f"  CB (x, z)           : ({self.cb_x_m:+.4f}, {self.cb_z_m:+.4f}) m",
            f"  BG (CB above CG)    : {self.bg_m * 1e3:+.1f} mm  -> {verdict}",
            f"  Righting stiffness  : {self.righting_stiffness_nm_per_rad:.3f} N·m/rad",
            f"  Long. CG-CB offset  : {self.longitudinal_offset_m * 1e3:+.1f} mm",
            f"  Natural pitch       : {self.natural_pitch_deg:+.2f} deg "
            "(at rest, from any built-in offset)",
        ]
        return "\n".join(lines)


def analyze(cg_x_m: float, cg_z_m: float,
            cb_x_m: float, cb_z_m: float,
            total_mass_kg: float,
            buoyancy_force_n: float | None = None,
            g: float = constants.GRAVITY) -> StabilityResult:
    """Static stability from CG and CB positions.

    ``buoyancy_force_n`` defaults to the vehicle weight (neutral buoyancy),
    which sets the magnitude of the righting couple.
    """
    bg = cb_z_m - cg_z_m  # positive when CB is above CG -> stable
    b = buoyancy_force_n if buoyancy_force_n is not None else total_mass_kg * g
    stiffness = b * bg  # dM/dtheta at theta=0 for M = B*BG*sin(theta)
    long_offset = cg_x_m - cb_x_m
    # At rest the vehicle pitches until CG is below CB: tan(theta)=offset/BG.
    natural_pitch = math.degrees(math.atan2(long_offset, bg)) if bg != 0 else 0.0
    return StabilityResult(
        total_mass_kg=total_mass_kg,
        cg_x_m=cg_x_m,
        cg_z_m=cg_z_m,
        cb_x_m=cb_x_m,
        cb_z_m=cb_z_m,
        bg_m=bg,
        is_stable=bg > 0,
        righting_stiffness_nm_per_rad=stiffness,
        longitudinal_offset_m=long_offset,
        natural_pitch_deg=natural_pitch,
    )


def pitch_from_mass_shift(mass_shift_kg: float,
                          shift_distance_m: float,
                          total_mass_kg: float,
                          bg_m: float) -> float:
    """Pitch angle (deg) produced by sliding ``mass_shift_kg`` by a distance.

    Positive ``shift_distance_m`` (forward) gives a nose-down pitch.
    """
    if bg_m <= 0:
        raise ValueError("bg_m must be positive (vehicle must be statically stable)")
    dx_cg = mass_shift_kg * shift_distance_m / total_mass_kg
    return math.degrees(math.atan2(dx_cg, bg_m))


def mass_shift_for_pitch(target_pitch_deg: float,
                         movable_mass_kg: float,
                         total_mass_kg: float,
                         bg_m: float) -> float:
    """Travel distance (m) of ``movable_mass_kg`` needed for a target pitch."""
    if bg_m <= 0:
        raise ValueError("bg_m must be positive (vehicle must be statically stable)")
    if movable_mass_kg <= 0:
        raise ValueError("movable_mass_kg must be positive")
    dx_cg = bg_m * math.tan(math.radians(target_pitch_deg))
    return dx_cg * total_mass_kg / movable_mass_kg

File: gliderforge/src/lib/calc/stability.ts
import { qty, type Quantity } from "@/lib/units";
import { G_STANDARD } from "@/lib/reference/water";
import { err, info, requireFinite, warn, type CalcResult, type CalcWarning } from "./types";
import type { Vec3 } from "./massprops";

/**
 * Static stability and trim of a FULLY SUBMERGED body.
 *
 * A submerged vehicle has no waterplane, so there is no metacentre in the
 * surface-ship sense. Its entire static righting moment comes from the
 * separation between the centre of buoyancy and the centre of gravity:
 *
 *   M_righting = rho * g * V * BG * sin(theta)
 *
 * where BG is the distance between CB and CG. The vehicle is statically
 * stable in pitch and roll if and only if the CB lies ABOVE the CG.
 *
 * Sign conventions used throughout (vehicle frame: +x forward, +y to port,
 * +z up):
 *   pitch  - nose-UP positive
 *   roll   - port-side-UP positive
 */

export interface StabilityInput {
  cg: Vec3;
  cb: Vec3;
  displacedVolumeSI: number;
  waterDensitySI: number;
  massSI: number;
  gravitySI?: number;
  /** Angle at which to report the righting moment, radians. */
  evaluationAngleSI?: number;
}

export interface StabilityValues extends Record<string, Quantity | undefined> {
  bgDistance: Quantity;
  verticalSeparation: Quantity;
  longitudinalSeparation: Quantity;
  lateralSeparation: Quantity;
  equilibriumPitch: Quantity;
  equilibriumRoll: Quantity;
  rightingMomentAtAngle: Quantity;
  rightingMomentPerRadian: Quantity;
  pitchStiffness: Quantity;
}

export function computeStability(input: StabilityInput): CalcResult<StabilityValues> {
  const warnings: CalcWarning[] = [];
  const g = input.gravitySI ?? G_STANDARD;
  const theta = input.evaluationAngleSI ?? (10 * Math.PI) / 180;

  const finite =
    Number.isFinite(input.cg.x) &&
    Number.isFinite(input.cg.z) &&
    Number.isFinite(input.cb.x) &&
    Number.isFinite(input.cb.z);
  if (!finite) {
    warnings.push(
      err(
        "The centre of gravity or centre of buoyancy could not be determined (missing component masses or volumes). No stability conclusion is possible from the data entered.",
      ),
    );
  }
  requireFinite(warnings, "Displaced volume", input.displacedVolumeSI, { min: 0, allowZero: true, field: "displacedVolumeSI" });

  const dx = input.cb.x - input.cg.x;
  const dy = input.cb.y - input.cg.y;
  const dz = input.cb.z - input.cg.z;
  const bg = Math.sqrt(dx * dx + dy * dy + dz * dz);

  const buoyantForce = input.waterDensitySI * g * input.displacedVolumeSI;
  const rightingAtAngle = buoyantForce * bg * Math.sin(theta);
  const stiffness = buoyantForce * bg; // dM/dtheta at small angles

  // Equilibrium attitude: the vehicle rotates until CB is directly above CG.
  const equilibriumPitch = Number.isFinite(dz) && dz !== 0 ? Math.atan2(dx, dz) : NaN;
  const equilibriumRoll = Number.isFinite(dz) && dz !== 0 ? Math.atan2(dy, dz) : NaN;

  if (finite) {
    if (dz <= 0) {
      warnings.push(
        err(
          `The centre of buoyancy is ${dz === 0 ? "at the same height as" : "BELOW"} the centre of gravity (dz = ${(dz * 1000).toFixed(1)} mm). A fully submerged vehicle in this condition has no righting moment in roll and will capsize or hang at an arbitrary attitude. Move mass down or displacement up.`,
        ),
      );
    } else if (dz < 0.002) {
      warnings.push(
        warn(
          `Vertical CB-CG separation is only ${(dz * 1000).toFixed(1)} mm. That is inside the build tolerance of most student vehicles, so the sign of the separation is not reliably known. Aim for at least 5-10 mm on a vehicle of this scale, and verify by watching how briskly it self-rights in a tank.`,
        ),
      );
    }
    if (Math.abs(dy) > 0.002) {
      warnings.push(
        warn(
          `Lateral CB-CG offset is ${(dy * 1000).toFixed(1)} mm, so the vehicle will hang with a permanent roll of about ${((equilibriumRoll * 180) / Math.PI).toFixed(1)} degrees. A rolled glider turns instead of gliding straight. Move mass laterally to null this.`,
        ),
      );
    }
    if (Math.abs(equilibriumPitch) > (30 * Math.PI) / 180) {
      warnings.push(
        warn(
          `The static equilibrium pitch of ${((equilibriumPitch * 180) / Math.PI).toFixed(1)} degrees is large. At this attitude the hull is far from aligned with the flow and the drag model's assumptions break down.`,
        ),
      );
    }
    const weight = input.massSI * g;
    if (Math.abs(buoyantForce - weight) / Math.max(weight, 1e-9) > 0.05) {
      warnings.push(
        info(
          "The vehicle is not close to neutral, so in practice it will accelerate vertically rather than sit at this static equilibrium attitude. The angles here describe the attitude it tends towards, not a hovering condition.",
        ),
      );
    }
  }
  warnings.push(
    info(
      "This is STATIC stability only. It says nothing about dynamic pitch oscillation, which depends on added mass, damping from the wings and fins, and how fast the buoyancy engine moves. A statically stable glider can still porpoise.",
    ),
  );

  return {
    id: "stability.static",
    title: "Static stability and trim",
    values: {
      bgDistance: qty(bg, "m"),
      verticalSeparation: qty(dz, "m"),
      longitudinalSeparation: qty(dx, "m"),
      lateralSeparation: qty(dy, "m"),
      equilibriumPitch: qty(equilibriumPitch, "rad"),
      equilibriumRoll: qty(equilibriumRoll, "rad"),
      rightingMomentAtAngle: qty(rightingAtAngle, "N*m"),
      rightingMomentPerRadian: qty(stiffness, "N*m"),
      pitchStiffness: qty(stiffness, "N*m"),
    },
    steps: [
      {
        label: "CB - CG vector",
        equation: "d = r_CB - r_CG",
        result: `(${(dx * 1000).toFixed(2)}, ${(dy * 1000).toFixed(2)}, ${(dz * 1000).toFixed(2)}) mm, |BG| = ${(bg * 1000).toFixed(2)} mm`,
      },
      {
        label: "Static equilibrium pitch",
        equation: "theta_eq = atan2(dx, dz)   [nose-up positive]",
        substitution: `atan2(${(dx * 1000).toFixed(2)} mm, ${(dz * 1000).toFixed(2)} mm)`,
        result: `${((equilibriumPitch * 180) / Math.PI).toFixed(2)} deg`,
        note: "The vehicle rotates until the CB sits directly above the CG. A CB aft of the CG gives a nose-down attitude.",
      },
      {
        label: "Static equilibrium roll",
        equation: "phi_eq = atan2(dy, dz)   [port-up positive]",
        result: `${((equilibriumRoll * 180) / Math.PI).toFixed(2)} deg`,
      },
      {
        label: "Righting moment",
        equation: "M = rho * g * V * BG * sin(theta)",
        substitution: `M = ${buoyantForce.toPrecision(4)} N * ${(bg * 1000).toFixed(2)} mm * sin(${((theta * 180) / Math.PI).toFixed(1)} deg)`,
        result: `${rightingAtAngle.toPrecision(4)} N*m at ${((theta * 180) / Math.PI).toFixed(1)} deg of disturbance`,
      },
      {
        label: "Small-angle stiffness",
        equation: "dM/dtheta = rho*g*V*BG",
        result: `${stiffness.toPrecision(4)} N*m per radian`,
      },
    ],
    equations: ["M = rho*g*V*BG*sin(theta)", "theta_eq = atan2(dx, dz)"],
    assumptions: [
      { text: "The vehicle is fully and continuously submerged.", basis: "There is no waterplane and therefore no metacentric contribution. On the surface this analysis does not apply." },
      { text: "The body is rigid and the CB does not move as the vehicle rotates.", basis: "True for a rigid hull. NOT true if a flexible bladder or trapped air pocket shifts with attitude." },
      { text: "No hydrodynamic moments.", basis: "Static analysis only. In motion, fins and the hull generate pitching moments that can dominate the buoyant righting moment." },
    ],
    warnings,
    inputs: input,
    confidence: finite ? "medium" : "low",
    limitations: [
      "Static only — no dynamics, no damping, no added mass, no hydrodynamic pitching moment.",
      "The result is only as good as the CB estimate, which depends on having every water-exposed volume entered correctly.",
    ],
  };
}

/* ------------------------------------------------------------------ */
/* Trim solver                                                         */
/* ------------------------------------------------------------------ */

export interface MovableComponent {
  id: string;
  name: string;
  massSI: number;
  positionX: number;
  /** Allowed travel along x, m. */
  minX: number;
  maxX: number;
}

export interface TrimSolution {
  componentId: string;
  componentName: string;
  currentX: number;
  requiredX: number;
  travelSI: number;
  withinLimits: boolean;
  limitedTo?: number;
  achievablePitchDeg: number;
  note: string;
}

export interface TrimSolverResult {
  targetPitchDeg: number;
  requiredCgShiftSI: number;
  requiredCgX: number;
  currentCgX: number;
  solutions: TrimSolution[];
  warnings: CalcWarning[];
  steps: { label: string; equation?: string; result: string }[];
  nonUniqueness: string;
}

/**
 * Solve for the component movement needed to reach a target static pitch.
 *
 * From the equilibrium condition tan(theta) = dx / dz with dx = x_CB - x_CG:
 *
 *   x_CG,required = x_CB - dz * tan(theta_target)
 *
 * Moving a single component of mass m by delta shifts the whole-vehicle CG by
 * m*delta / M_total, so
 *
 *   delta = (x_CG,required - x_CG,current) * M_total / m
 *
 * Each movable component is solved INDEPENDENTLY. Any combination of these
 * movements that produces the same total moment shift also works, which is
 * why the result is explicitly reported as non-unique.
 */
export function solveTrim(
  targetPitchRad: number,
  current: { cgX: number; cbX: number; verticalSeparationSI: number; totalMassSI: number },
  movable: MovableComponent[],
): TrimSolverResult {
  const warnings: CalcWarning[] = [];
  const dz = current.verticalSeparationSI;

  if (!Number.isFinite(dz) || dz <= 0) {
    warnings.push(
      err(
        "The vertical CB-CG separation is zero, negative or unknown. Pitch trim cannot be solved: with the CB at or below the CG there is no restoring moment to trim against.",
      ),
    );
  }
  const requiredCgX = current.cbX - dz * Math.tan(targetPitchRad);
  const shift = requiredCgX - current.cgX;

  const solutions: TrimSolution[] = movable.map((c) => {
    const delta = c.massSI > 0 ? (shift * current.totalMassSI) / c.massSI : NaN;
    const requiredX = c.positionX + delta;
    const within = requiredX >= c.minX - 1e-12 && requiredX <= c.maxX + 1e-12;
    const limited = within ? undefined : Math.min(c.maxX, Math.max(c.minX, requiredX));
    // Pitch actually achievable if the component is driven to its limit.
    const achievableCgX =
      limited === undefined ? requiredCgX : current.cgX + ((limited - c.positionX) * c.massSI) / current.totalMassSI;
    const achievablePitch = Math.atan2(current.cbX - achievableCgX, dz);
    return {
      componentId: c.id,
      componentName: c.name,
      currentX: c.positionX,
      requiredX,
      travelSI: delta,
      withinLimits: within,
      limitedTo: limited,
      achievablePitchDeg: (achievablePitch * 180) / Math.PI,
      note: within
        ? `Move "${c.name}" by ${(delta * 1000).toFixed(1)} mm ${delta >= 0 ? "forward" : "aft"}.`
        : `"${c.name}" would have to move to x = ${(requiredX * 1000).toFixed(1)} mm, outside its allowed travel of ${(c.minX * 1000).toFixed(0)}..${(c.maxX * 1000).toFixed(0)} mm. At its limit it can only reach ${((achievablePitch * 180) / Math.PI).toFixed(1)} deg.`,
    };
  });

  if (movable.length === 0) {
    warnings.push(
      err("No components were marked movable, so there is nothing to trim with. Mark the battery or a ballast mass as movable and give it a travel range."),
    );
  }
  if (solutions.length > 0 && solutions.every((s) => !s.withinLimits)) {
    warnings.push(
      err("No single movable component can reach the target pitch within its travel limits. Either combine several movements, add a dedicated trim mass, or relax the target."),
    );
  }
  for (const s of solutions) {
    if (Math.abs(s.travelSI) > 0.5) {
      warnings.push(
        warn(`The movement required of "${s.componentName}" (${(s.travelSI * 1000).toFixed(0)} mm) is very large — it probably exceeds the vehicle. Consider a heavier trim mass instead.`),
      );
    }
  }
  warnings.push(
    warn(
      "Moving a component changes the displaced-volume distribution too if that component is water-exposed, which moves the CB as well as the CG. This solver holds the CB fixed; re-run the mass-properties calculation after applying a move and check the result.",
    ),
  );

  return {
    targetPitchDeg: (targetPitchRad * 180) / Math.PI,
    requiredCgShiftSI: shift,
    requiredCgX,
    currentCgX: current.cgX,
    solutions,
    warnings,
    steps: [
      {
        label: "Required CG longitudinal position",
        equation: "x_CG,req = x_CB - dz * tan(theta_target)",
        result: `${(requiredCgX * 1000).toFixed(2)} mm (currently ${(current.cgX * 1000).toFixed(2)} mm)`,
      },
      {
        label: "Required CG shift",
        equation: "shift = x_CG,req - x_CG,now",
        result: `${(shift * 1000).toFixed(2)} mm ${shift >= 0 ? "forward" : "aft"}`,
      },
      {
        label: "Movement of a single component",
        equation: "delta = shift * M_total / m_component",
        result: `${movable.length} candidate component(s) evaluated independently`,
      },
    ],
    nonUniqueness:
      "This is NOT a unique answer. Any set of component movements whose total moment change equals shift x M_total achieves the same trim — moving one heavy item a little, or a light item a lot, or several items together. The options listed are single-component solutions offered as starting points, not a prescription.",
  };
}
